File: cpp/p1745.cpp
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    bool checkPartitioning(string s) {
        int n = s.size();
        vector<vector<bool>> f(n, vector<bool>(n, false));
        for (int i = n - 1; i >= 0; i--) {
            f[i][i] = true;
            for (int j = i + 1; j < n; j++) {
                f[i][j] = s[i] == s[j] ? (j - i < 2 || f[i + 1][j - 1]) : false;
            }
        }
        vector<vector<bool>> g(3, vector<bool>(n, false));
        g[0][0] = true;
        for (int i = 0; i < 3; i++) {
            for (int j = 1; j < n; j++) {
                if (i == 0) {
                    g[i][j] = f[0][j];
                } else {
                    for (int k = 1; k <= j; k++) {
                        g[i][j] = g[i][j] || (g[i - 1][k - 1] && f[k][j]);
                    }
                }
            }
        }
        return g[2][n - 1];
    }
};

int main() {
    assert(Solution().checkPartitioning("abcbdd"));
    assert(!Solution().checkPartitioning("bcbddxy"));
    return 0;
}

File: rust/src/problems/p1745.rs
pub struct Solution;

impl Solution {
    pub fn check_partitioning(s: String) -> bool {
        let (n, s) = (s.len(), s.as_bytes());
        let mut f = vec![vec![false; n]; n];
        for i in (0..n).rev() {
            f[i][i] = true;
            for j in i + 1..n {
                f[i][j] = if s[i] == s[j] {
                    j - i == 1 || f[i + 1][j - 1]
                } else {
                    false
                };
            }
        }
        for i in 0..n - 2 {
            if !f[0][i] {
                continue;
            }
            for j in i + 1..n - 1 {
                if f[i + 1][j] && f[j + 1][n - 1] {
                    return true;
                }
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        assert!(Solution::check_partitioning("abcbdd".to_string()));
    }

    #[test]
    fn case2() {
        assert!(!Solution::check_partitioning("bcbddxy".to_string()));
    }
}
